File: Python/_117_Populating_Next_Right_Pointers_in_Each_Node_II/solution.py
# Definition for binary tree with next pointer.
# class TreeLinkNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None

class Solution:
    # @param root, a tree link node
    # @return nothing
    def connect(self, root):
        tmpRoot = root
        while tmpRoot:
            cur = tmpRoot
            dummy = TreeLinkNode(0)
            tmp = dummy
            while cur:
                if cur.left:
                    tmp.next = cur.left
                    tmp = tmp.next
                if cur.right:
                    tmp.next = cur.right
                    tmp = tmp.next
                cur = cur.next      
            tmpRoot = dummy.next

File: Java/_117_Populating_Next_Right_Pointers_in_Each_Node_II/Solution.java
/**
 * Definition for binary tree with next pointer.
 * public class TreeLinkNode {
 *     int val;
 *     TreeLinkNode left, right, next;
 *     TreeLinkNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void connect(TreeLinkNode root) {
        TreeLinkNode tmpRoot = root;    
        while (tmpRoot != null) {
            TreeLinkNode cur = tmpRoot;
            TreeLinkNode dummy = new TreeLinkNode(0);
            TreeLinkNode tmp = dummy;
            while (cur != null) {
                if (cur.left != null) {
                    tmp.next = cur.left;
                    tmp = tmp.next;
                }
                if (cur.right != null) {
                    tmp.next = cur.right;
                    tmp = tmp.next;
                }
                cur = cur.next;
            }          
            tmpRoot = dummy.next;
        }
    }
}